tt_metal/impl/dispatch/arch.cpp

```diff
@@ -0,0 +1,738 @@
+            uint32_t num_routers = device_id_to_num_routers[router_kernel->GetDeviceId()];
+
+            // Special case for MMIO chips, can have routers servicing different tunnels
+            chip_id_t mmio_device_id = tt::Cluster::instance().get_associated_mmio_device(router_kernel->GetDeviceId());
+            if (router_kernel->GetDeviceId() == mmio_device_id) {
+                uint32_t num_tunnels = tt::Cluster::instance().get_tunnels_from_mmio_device(mmio_device_id).size();
+                num_routers /= num_tunnels;
+                uint32_t router_vcs =  (router_vcs_for_device + num_routers - 1) / num_routers;
+                router_kernel->SetVCCount(router_vcs);
+            } else {
+                uint32_t router_vcs = (router_vcs_for_device + remaining_routers - 1) / num_routers;
+                router_kernel->SetVCCount(router_vcs);
+            }
+            device_id_to_remaining_routers[router_kernel->GetDeviceId()]--;
+        }
+    }
+}
+
+std::unique_ptr<Program> create_cq_program(Device *device) {
+    TT_ASSERT(
+        node_id_to_kernel.size() > 0,
+        "Tried to create CQ program without nodes populated (need to run populate_fd_kernels()");
+
+    // First pass, add device/program to all kernels for this device and generate static configs.
+    auto cq_program_ptr = std::make_unique<Program>();
+    // for (auto &node_and_kernel : node_id_to_kernel) {
+    for (int idx = 0; idx < node_id_to_kernel.size(); idx++) {
+        if (node_id_to_kernel[idx]->GetDeviceId() == device->id()) {
+            node_id_to_kernel[idx]->AddDeviceAndProgram(device, cq_program_ptr.get());
+            tt::log_debug("GenerateStaticConfigs for Node {}", idx);
+            node_id_to_kernel[idx]->GenerateStaticConfigs();
+            tt::log_debug("Node {} has coord: {} (phys={})", idx, node_id_to_kernel[idx]->GetLogicalCore().str(), node_id_to_kernel[idx]->GetVirtualCore().str());
+        }
+    }
+
+    // Third pass, populate dependent configs and create kernels for each node
+    // for (auto &node_and_kernel : node_id_to_kernel) {
+    for (int idx = 0; idx < node_id_to_kernel.size(); idx++) {
+        if (node_id_to_kernel[idx]->GetDeviceId() == device->id()) {
+            tt::log_debug("GenerateDependentConfigs for Node {}", idx);
+            node_id_to_kernel[idx]->GenerateDependentConfigs();
+            tt::log_debug("CreateKernel for Node {}", idx);
+            node_id_to_kernel[idx]->CreateKernel();
+        }
+    }
+
+    // Compile the program and return it so Device can register it
+    detail::CompileProgram(device, *cq_program_ptr, /*fd_bootloader_mode=*/true);
+    tt::log_debug("Done Compiling CQ Program for Device {}, grid size = {}", device->id(), device->compute_with_storage_grid_size().str());
+    return cq_program_ptr;
+}
+
+void configure_dispatch_cores(Device *device) {
+    // Set up completion_queue_writer core. This doesn't actually have a kernel so keep it out of the struct and config
```

Comment:
doesn't this have a kernel on it today e.g. for single device single cq dispatch it would be `cq_dispatch.cpp` that writes to completion queue

Reply:
Yeah I was a bit confused about this core, it looks like it's always the same as the DISPATCH_H for a given device? Is that always going to be the same going forwards? If so, wondering if this should just be collapsed to use the DISPATCH_H core assignment